api/app/src/mappings/google/sleep.ts

```diff
@@ -17,16 +23,36 @@ export const mapToSleep = (googleSleep: GoogleSleep, date: string): Sleep => {
 
   googleSleep.bucket[0].dataset.forEach(data => {
     if (data.point.length) {
-      const startTimeNanos = Number(data.point[0].startTimeNanos);
-      const endTimeNanos = Number(data.point[0].endTimeNanos);
+      const sortedDates = sortBy(data.point, "startTimeNanos");
+
+      const sleepSession = sortedDates.reduce((acc: SleepSession[], curr) => {
+        if (acc.length) {
+          const lastItem = acc[acc.length - 1];
+          const currDate = dayjs(convert(Number(curr.startTimeNanos)).from("ns").to("ms"));
+
+          const timeDifference = currDate.diff(lastItem.endTime, "minutes");
+
+          if (timeDifference >= 30) {
+            return acc;
+          }
+        }
+
+        const startTimeNanos = Number(curr.startTimeNanos);
+        const endTimeNanos = Number(curr.endTimeNanos);
+
+        acc.push({
+          startTime: dayjs(convert(startTimeNanos).from("ns").to("ms")).format(
+            "YYYY-MM-DDTHH:mm:ssZ"
+          ),
+          endTime: dayjs(convert(endTimeNanos).from("ns").to("ms")).format("YYYY-MM-DDTHH:mm:ssZ"),
+        });
+
+        return acc;
+      }, []);
 
       if (data.dataSourceId === sourceIdSleep) {
-        sleep.start_time = dayjs(convert(startTimeNanos).from("ns").to("ms")).format(
-          "YYYY-MM-DDTHH:mm:ssZ"
-        );
-        sleep.end_time = dayjs(convert(endTimeNanos).from("ns").to("ms")).format(
-          "YYYY-MM-DDTHH:mm:ssZ"
-        );
+        sleep.start_time = sleepSession[0].startTime;
+        sleep.end_time = sleepSession[sleepSession.length - 1].endTime;
```

Comment:
Are we sure there will be elements on this array? Same for the `forEach` that wraps this code.

Reply:
Yea if its empty it will return this 
```
{
  startTimeMillis: '1679270400000',
  endTimeMillis: '1679356800000',
  dataset: [
    {
      dataSourceId: 'derived:com.google.sleep.segment:com.google.android.gms:merged',
      point: []
    }
  ]
} 
```